File: Cpp/Split_Array_Largest_Sum.cpp
class Solution {
public:
    int splitArray(vector<int>& nums, int k) {
        int low = 0, high = 0;
        for (int n : nums) {
            low = max(n, low);
            high += n;
        }
        int ans = high;
        while (low <= high) {
            long long mid = low + (high - low) / 2;
            int reqPartitions = countPartitions(nums, mid);
            if (reqPartitions > k)
                low = mid + 1;
            else {
                ans = mid;
                high = mid - 1;
            }
        }
        return ans;
    }

    int countPartitions(vector<int>& nums, long long maxAllowed) {
        int partitions = 1;
        long long currSum = 0;
        for (int n : nums) {
            if (currSum + n <= maxAllowed)
                currSum += n;
            else {
                partitions++;
                currSum = n;
            }
        }
        return partitions;
    }
};

File: Rust/Split_Array_Largest_Sum.rs
impl Solution {
    pub fn split_array(nums: Vec<i32>, k: i32) -> i32 {
        let mut low: i64 = *nums.iter().max().unwrap() as i64;
        let mut high: i64 = nums.iter().map(|&x| x as i64).sum();
        let mut ans = high;
        while low <= high {
            let mid = low + (high - low) / 2;

            if Self::count_partitions(&nums, mid) > k {
                low = mid + 1;
            } else {
                ans = mid;
                high = mid - 1;
            }
        }

        ans as i32
    }
    fn count_partitions(nums: &Vec<i32>, max_allowed: i64) -> i32 {
        let mut partitions = 1;
        let mut current_sum = 0;

        for &num in nums {
            if current_sum + (num as i64) <= max_allowed {
                current_sum += num as i64;
            } else {
                partitions += 1;
                current_sum = num as i64;
            }
        }
        partitions
    }
}
